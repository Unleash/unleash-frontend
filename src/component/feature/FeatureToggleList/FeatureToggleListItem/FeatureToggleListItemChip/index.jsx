import { connect } from 'react-redux';
import Component from './FeatureToggleListItemChip';

const mapStateToProps = state => ({
    types: state.featureTypes.toJS(),
});

const FeatureType = connect(mapStateToProps)(Component);

export default FeatureType;
