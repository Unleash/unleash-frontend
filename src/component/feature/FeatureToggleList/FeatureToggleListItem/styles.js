import { makeStyles } from "@material-ui/styles";

export const useStyles = makeStyles(theme => ({
    listItem: {
        padding: "0.25rem 0"
    },
    listItemMetric: {
        width: "40px",
        marginRight: "1rem",
        flexShrink: "0"
    },
    listItemSvg: {
        fill: theme.palette.icons.lightGrey
    },
    listItemLink: {
        marginLeft: "10px",
        minWidth: "0"
    },
    listItemStrategies: {
        marginLeft: "auto"
    }
}));
