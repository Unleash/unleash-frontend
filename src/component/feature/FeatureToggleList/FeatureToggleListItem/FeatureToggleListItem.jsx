import React, { memo } from "react";
import PropTypes from "prop-types";
import classnames from "classnames";

import { Link } from "react-router-dom";
import { Switch, Icon, IconButton, ListItem } from "@material-ui/core";
import TimeAgo from "react-timeago";
import Progress from "../../progress-component";
import Status from "../../status-component";
import FeatureToggleListItemChip from "./FeatureToggleListItemChip";
import ConditionallyRender from "../../../common/conditionally-render";

import { UPDATE_FEATURE } from "../../../../permissions";
import { calc, styles as commonStyles } from "../../../common";

import { useStyles } from "./styles";

const Feature = ({
    feature,
    toggleFeature,
    settings,
    metricsLastHour = { yes: 0, no: 0, isFallback: true },
    metricsLastMinute = { yes: 0, no: 0, isFallback: true },
    revive,
    hasPermission
}) => {
    const styles = useStyles();

    const { name, description, enabled, type, stale, createdAt } = feature;
    const { showLastHour = false } = settings;
    const isStale = showLastHour
        ? metricsLastHour.isFallback
        : metricsLastMinute.isFallback;
    const percent =
        1 *
        (showLastHour
            ? calc(
                  metricsLastHour.yes,
                  metricsLastHour.yes + metricsLastHour.no,
                  0
              )
            : calc(
                  metricsLastMinute.yes,
                  metricsLastMinute.yes + metricsLastMinute.no,
                  0
              ));
    const featureUrl =
        toggleFeature === undefined
            ? `/archive/strategies/${name}`
            : `/features/strategies/${name}`;

    return (
        <ListItem className={styles.listItem}>
            <span className={styles.listItemMetric}>
                <Progress
                    strokeWidth={15}
                    percentage={percent}
                    isFallback={isStale}
                />
            </span>
            <span className={styles.listItemToggle}>
                <ConditionallyRender
                    condition={hasPermission(UPDATE_FEATURE)}
                    show={
                        <Switch
                            disabled={toggleFeature === undefined}
                            title={`Toggle ${name}`}
                            key="left-actions"
                            onChange={() => toggleFeature(!enabled, name)}
                            checked={enabled}
                        />
                    }
                    elseShow={
                        <Switch
                            disabled
                            title={`Toggle ${name}`}
                            key="left-actions"
                            checked={enabled}
                        />
                    }
                />
            </span>
            <span className={classnames(styles.listItemLink)}>
                <Link
                    to={featureUrl}
                    className={classnames(
                        commonStyles.listLink,
                        commonStyles.truncate
                    )}
                >
                    <span className={commonStyles.toggleName}>
                        {name}&nbsp;
                    </span>
                    <small>
                        <TimeAgo date={createdAt} live={false} />
                    </small>
                    <div>
                        <span className={commonStyles.truncate}>
                            {description}
                        </span>
                    </div>
                </Link>
            </span>
            <span
                className={classnames(
                    styles.listItemStrategies,
                    commonStyles.hideLt920
                )}
            >
                <Status stale={stale} showActive={false} />
                <FeatureToggleListItemChip type={type} />
            </span>
            <ConditionallyRender
                condition={revive && hasPermission(UPDATE_FEATURE)}
                show={
                    <IconButton onClick={() => revive(feature.name)}>
                        <Icon>undo</Icon>
                    </IconButton>
                }
                elseShow={<span />}
            />
        </ListItem>
    );
};

Feature.propTypes = {
    feature: PropTypes.object,
    toggleFeature: PropTypes.func,
    settings: PropTypes.object,
    metricsLastHour: PropTypes.object,
    metricsLastMinute: PropTypes.object,
    revive: PropTypes.func,
    hasPermission: PropTypes.func.isRequired
};

export default memo(Feature);
