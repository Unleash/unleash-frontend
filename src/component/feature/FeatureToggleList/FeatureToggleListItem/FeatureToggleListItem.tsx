import { memo } from 'react';
import classnames from 'classnames';
import { Link } from 'react-router-dom';
import { Chip, ListItem, Tooltip } from '@material-ui/core';
import { Undo } from '@material-ui/icons';
import TimeAgo from 'react-timeago';
import { IAccessContext } from 'contexts/AccessContext';
import StatusChip from 'component/common/StatusChip/StatusChip';
import { ConditionallyRender } from 'component/common/ConditionallyRender/ConditionallyRender';
import { UPDATE_FEATURE } from 'component/providers/AccessProvider/permissions';
import { styles as commonStyles } from 'component/common'; // FIXME: remove
import { IFeatureToggle } from 'interfaces/featureToggle';
import { IFlags } from 'interfaces/uiConfig';
import { getTogglePath } from 'utils/routePathHelpers';
import FeatureStatus from 'component/feature/FeatureView/FeatureStatus/FeatureStatus';
import FeatureType from 'component/feature/FeatureView/FeatureType/FeatureType';
import useProjects from 'hooks/api/getters/useProjects/useProjects';
import PermissionIconButton from 'component/common/PermissionIconButton/PermissionIconButton';
import { useStyles } from './styles'; // FIXME: cleanup

interface IFeatureToggleListItemProps {
    feature: IFeatureToggle;
    onRevive?: (id: string) => void;
    hasAccess: IAccessContext['hasAccess'];
    flags?: IFlags;
    className?: string;
}

export const FeatureToggleListItem = memo<IFeatureToggleListItemProps>(
    ({ feature, onRevive, hasAccess, flags = {}, className, ...rest }) => {
        const styles = useStyles();

        const { projects } = useProjects();
        const isArchive = !!onRevive;

        const {
            name,
            description,
            type,
            stale,
            createdAt,
            project,
            lastSeenAt,
        } = feature;

        const projectExists = () => {
            let projectExist = projects.find(proj => proj.id === project);
            if (projectExist) {
                return true;
            }
            return false;
        };

        const reviveFeature = () => {
            if (projectExists() && onRevive) {
                onRevive(feature.name);
            }
        };

        return (
            <ListItem
                {...rest}
                className={classnames(styles.listItem, className)}
            >
                <span className={styles.listItemMetric}>
                    <FeatureStatus
                        lastSeenAt={lastSeenAt}
                        tooltipPlacement="left"
                    />
                </span>
                <span
                    className={classnames(
                        styles.listItemType,
                        commonStyles.hideLt600
                    )}
                >
                    <FeatureType type={type} />
                </span>
                <span className={classnames(styles.listItemLink)}>
                    <ConditionallyRender
                        condition={!isArchive}
                        show={
                            <Link
                                to={getTogglePath(feature.project, name)}
                                className={classnames(
                                    commonStyles.listLink,
                                    commonStyles.truncate
                                )}
                            >
                                <Tooltip title={description || ' '}>
                                    <span className={commonStyles.toggleName}>
                                        {name}&nbsp;
                                    </span>
                                </Tooltip>
                                {/* <span className={styles.listItemToggle}></span> */}
                                <span></span>
                                <small>
                                    <TimeAgo date={createdAt} live={false} />
                                </small>
                                <div>
                                    <span className={commonStyles.truncate}>
                                        <small>{description}</small>
                                    </span>
                                </div>
                            </Link>
                        }
                        elseShow={
                            <>
                                <Tooltip title={description || ' '}>
                                    <span className={commonStyles.toggleName}>
                                        {name}&nbsp;{' '}
                                    </span>
                                </Tooltip>
                                {/* <span className={styles.listItemToggle}></span> */}
                                <span></span>
                                <small>
                                    <TimeAgo date={createdAt} live={false} />
                                </small>
                                <div>
                                    <span className={commonStyles.truncate}>
                                        <small>{description}</small>
                                    </span>
                                </div>
                            </>
                        }
                    />
                </span>
                <span
                    className={classnames(
                        styles.listItemStrategies,
                        commonStyles.hideLt920
                    )}
                >
                    <StatusChip stale={stale} showActive={false} />
                    <Link
                        to={`/projects/${project}`}
                        style={{ textDecoration: 'none' }}
                        className={classnames({
                            [`${styles.disabledLink}`]: !projectExists(),
                        })}
                    >
                        <Chip
                            color="primary"
                            variant="outlined"
                            // className={styles.typeChip} // FIXME: cleanup
                            style={{ marginLeft: '8px', cursor: 'pointer' }}
                            title={`Project: ${project}`}
                            label={project}
                        />
                    </Link>
                </span>
                <ConditionallyRender
                    condition={isArchive}
                    show={
                        <PermissionIconButton
                            permission={UPDATE_FEATURE}
                            projectId={project}
                            disabled={
                                !hasAccess(UPDATE_FEATURE, project) ||
                                !projectExists()
                            }
                            onClick={reviveFeature}
                        >
                            <Undo />
                        </PermissionIconButton>
                    }
                />
            </ListItem>
        );
    }
);
