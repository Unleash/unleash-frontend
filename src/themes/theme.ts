import { createTheme } from '@mui/material/styles';
import { colors } from './colors';

export default createTheme({
    boxShadows: {
        main: '0px 2px 4px rgba(129, 122, 254, 0.2)',
    },
    typography: {
        fontFamily: 'Sen, Roboto, sans-serif',
        fontWeightBold: '700',
        fontWeightMedium: '700',
        allVariants: { lineHeight: 1.4 },
        button: { lineHeight: 1.75 },
        h1: {
            fontSize: '1.5rem',
            lineHeight: 1.875,
        },
    },
    fontSizes: {
        mainHeader: '1.25rem',
        subHeader: '1.1rem',
        bodySize: '1rem',
        smallBody: '0.9rem',
        smallerBody: '0.8rem',
    },
    fontWeight: {
        thin: 300,
        medium: 400,
        semi: 700,
        bold: 700,
    },
    shape: {
        borderRadius: '3px',
        borderRadiusLarge: '10px',
        borderRadiusExtraLarge: '25px',
    },
    palette: {
        primary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        secondary: {
            main: colors.purple[800],
            light: colors.purple[700],
            dark: colors.purple[900],
        },
        info: {
            light: colors.blue[700],
            main: colors.blue[700],
            dark: colors.blue[800],
        },
        success: {
            light: colors.green[700],
            main: colors.green[700],
            dark: colors.green[800],
        },
        warning: {
            light: colors.orange[700],
            main: colors.orange[700],
            dark: colors.orange[800],
        },
        error: {
            light: colors.red[700],
            main: colors.red[700],
            dark: colors.red[800],
        },
        divider: colors.grey[300],
        dividerAlternative: colors.grey[500],
        grey: colors.grey,
        text: {
            primary: colors.grey[900],
            secondary: colors.grey[800],
            disabled: colors.grey[600],
        },
        code: {
            main: '#0b8c8f',
            diffAdd: 'green',
            diffSub: 'red',
            diffNeutral: 'black',
            edited: 'blue',
            background: '#efefef',
        },
        activityIndicators: {
            unknown: colors.grey[100],
            recent: colors.green[100],
            inactive: colors.orange[200],
            abandoned: colors.red[200],
        },
    },
    components: {
        MuiTableHead: {
            styleOverrides: {
                root: {
                    background: colors.grey[200],
                },
            },
        },
        MuiLink: {
            styleOverrides: {
                root: {
                    color: colors.purple[900],
                },
            },
        },
        MuiBreadcrumbs: {
            styleOverrides: {
                root: {
                    '& a': {
                        color: colors.purple[900],
                    },
                },
            },
        },
        MuiTableRow: {
            styleOverrides: {
                root: {
                    '&.MuiTableRow-hover:hover': {
                        background: colors.grey[100],
                    },
                },
            },
        },
    },
});
