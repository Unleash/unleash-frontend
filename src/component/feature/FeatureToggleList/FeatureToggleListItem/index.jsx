import FeatureToggleListItem from "./FeatureToggleListItem";

export default FeatureToggleListItem;
