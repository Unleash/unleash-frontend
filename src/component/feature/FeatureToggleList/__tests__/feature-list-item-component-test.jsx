import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { ThemeProvider } from '@material-ui/core';

import FeatureToggleListItem from '../FeatureToggleListItem';
import renderer from 'react-test-renderer';

import theme from '../../../../themes/main-theme';

jest.mock('../FeatureToggleListItem/FeatureToggleListItemChip');

test('renders correctly with one feature', () => {
    const feature = {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        strategies: [
            {
                name: 'gradualRolloutRandom',
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
    };
    const featureMetrics = { lastHour: {}, lastMinute: {}, seenApps: {} };
    const settings = { sort: 'name' };
    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <FeatureToggleListItem
                    key={0}
                    settings={settings}
                    metricsLastHour={featureMetrics.lastHour[feature.name]}
                    metricsLastMinute={featureMetrics.lastMinute[feature.name]}
                    feature={feature}
                    toggleFeature={jest.fn()}
                    hasAccess={() => true}
                />
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});

test('renders correctly with one feature without permission', () => {
    const feature = {
        name: 'Another',
        description: "another's description",
        enabled: false,
        stale: false,
        strategies: [
            {
                name: 'gradualRolloutRandom',
                parameters: {
                    percentage: 50,
                },
            },
        ],
        createdAt: '2018-02-04T20:27:52.127Z',
    };
    const featureMetrics = { lastHour: {}, lastMinute: {}, seenApps: {} };
    const settings = { sort: 'name' };
    const tree = renderer.create(
        <MemoryRouter>
            <ThemeProvider theme={theme}>
                <FeatureToggleListItem
                    key={0}
                    settings={settings}
                    metricsLastHour={featureMetrics.lastHour[feature.name]}
                    metricsLastMinute={featureMetrics.lastMinute[feature.name]}
                    feature={feature}
                    toggleFeature={jest.fn()}
                    hasAccess={() => true}
                />
            </ThemeProvider>
        </MemoryRouter>
    );

    expect(tree).toMatchSnapshot();
});
