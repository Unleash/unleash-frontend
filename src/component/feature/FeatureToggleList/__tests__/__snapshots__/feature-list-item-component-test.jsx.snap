// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with one feature 1`] = `
<li
  className="MuiListItem-root makeStyles-listItem-1 MuiListItem-gutters"
  disabled={false}
>
  <span
    className="makeStyles-listItemMetric-2"
  >
    <div
      aria-describedby={null}
      className="makeStyles-container-7"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "background": "#EDF0F1",
          "fontSize": "0.8rem",
        }
      }
      title="No usage reported from connected applications"
    >
      <span
        style={
          Object {
            "fontSize": "1.4rem",
          }
        }
      >
        ⊕
      </span>
    </div>
  </span>
  <span
    className="makeStyles-listItemType-3 hideLt600"
  >
    <svg
      aria-describedby={null}
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-8"
      data-loading={true}
      focusable="false"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title={null}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
      />
    </svg>
  </span>
  <span
    className="makeStyles-listItemLink-5"
  >
    <a
      className="listLink truncate"
      href="/projects/default/features/Another/strategies"
      onClick={[Function]}
    >
      <span
        aria-describedby={null}
        className="toggleName"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="another's description"
      >
        Another
         
      </span>
      <span />
      <small>
        <time
          dateTime="2018-02-04T20:27:52.127Z"
          title="2018-02-04T20:27:52.127Z"
        >
          4 years ago
        </time>
      </small>
      <div>
        <span
          className="truncate"
        >
          <small>
            another's description
          </small>
        </span>
      </div>
    </a>
  </span>
  <span
    className="makeStyles-listItemStrategies-6 hideLt920"
  >
    <a
      href="/projects/default"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      <div
        className="MuiChip-root MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        style={
          Object {
            "marginLeft": "8px",
          }
        }
        title="Project: default"
      >
        <span
          className="MuiChip-label"
        >
          default
        </span>
      </div>
    </a>
  </span>
</li>
`;

exports[`renders correctly with one feature without permission 1`] = `
<li
  className="MuiListItem-root makeStyles-listItem-1 MuiListItem-gutters"
  disabled={false}
>
  <span
    className="makeStyles-listItemMetric-2"
  >
    <div
      aria-describedby={null}
      className="makeStyles-container-7"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      style={
        Object {
          "background": "#EDF0F1",
          "fontSize": "0.8rem",
        }
      }
      title="No usage reported from connected applications"
    >
      <span
        style={
          Object {
            "fontSize": "1.4rem",
          }
        }
      >
        ⊕
      </span>
    </div>
  </span>
  <span
    className="makeStyles-listItemType-3 hideLt600"
  >
    <svg
      aria-describedby={null}
      aria-hidden={true}
      className="MuiSvgIcon-root makeStyles-icon-8"
      data-loading={true}
      focusable="false"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseLeave={[Function]}
      onMouseOver={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      title={null}
      viewBox="0 0 24 24"
    >
      <path
        d="M12 4V1L8 5l4 4V6c3.31 0 6 2.69 6 6 0 1.01-.25 1.97-.7 2.8l1.46 1.46C19.54 15.03 20 13.57 20 12c0-4.42-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6 0-1.01.25-1.97.7-2.8L5.24 7.74C4.46 8.97 4 10.43 4 12c0 4.42 3.58 8 8 8v3l4-4-4-4v3z"
      />
    </svg>
  </span>
  <span
    className="makeStyles-listItemLink-5"
  >
    <a
      className="listLink truncate"
      href="/projects/undefined/features/Another/strategies"
      onClick={[Function]}
    >
      <span
        aria-describedby={null}
        className="toggleName"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="another's description"
      >
        Another
         
      </span>
      <span />
      <small>
        <time
          dateTime="2018-02-04T20:27:52.127Z"
          title="2018-02-04T20:27:52.127Z"
        >
          4 years ago
        </time>
      </small>
      <div>
        <span
          className="truncate"
        >
          <small>
            another's description
          </small>
        </span>
      </div>
    </a>
  </span>
  <span
    className="makeStyles-listItemStrategies-6 hideLt920"
  >
    <a
      href="/projects/undefined"
      onClick={[Function]}
      style={
        Object {
          "textDecoration": "none",
        }
      }
    >
      <div
        className="MuiChip-root MuiChip-colorPrimary MuiChip-outlined MuiChip-outlinedPrimary"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        style={
          Object {
            "marginLeft": "8px",
          }
        }
        title="Project: undefined"
      >
        <span
          className="MuiChip-label"
        />
      </div>
    </a>
  </span>
</li>
`;
